File: src/main/java/co/gatedaccess/web/model/JoinCommunityRequest.java
package co.gatedaccess.web.model;

import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import java.util.Date;

@Document("join_community_request")
public class JoinCommunityRequest {
    @Id
    String id;
    @DBRef
    Member member;
    @DBRef
    Community community;
    @Field("created_at")
    @CreatedDate
    Date createdAt;
    @Field("accepted_at")
    Date acceptedAt;
    @Field("rejected_at")
    Date rejectAt;
    @DBRef
    Member referrer;

    public String getId() {
        return id;
    }

    public Member getMember() {
        return member;
    }

    public Community getCommunity() {
        return community;
    }

    public Date getCreatedAt() {
        return createdAt;
    }

    public Date getAcceptedAt() {
        return acceptedAt;
    }

    public Date getRejectAt() {
        return rejectAt;
    }

    public void setAcceptedAt(Date acceptedAt) {
        this.acceptedAt = acceptedAt;
    }

    public void setRejectAt(Date rejectAt) {
        this.rejectAt = rejectAt;
    }

    public static final class Builder {
        private String id;
        private Member member;
        private Community community;
        private Date createdAt;
        private Date acceptedAt;
        private Date rejectAt;
        private Member referrer;

        public Builder() {
        }

        public static Builder aJoinCommunityRequest() {
            return new Builder();
        }

        public Builder withId(String id) {
            this.id = id;
            return this;
        }

        public Builder withMember(Member member) {
            this.member = member;
            return this;
        }

        public Builder withCommunity(Community community) {
            this.community = community;
            return this;
        }

        public Builder withCreatedAt(Date createdAt) {
            this.createdAt = createdAt;
            return this;
        }

        public Builder withAcceptedAt(Date acceptedAt) {
            this.acceptedAt = acceptedAt;
            return this;
        }

        public Builder withRejectAt(Date rejectAt) {
            this.rejectAt = rejectAt;
            return this;
        }

        public Builder withReferrer(Member referrer) {
            this.referrer = referrer;
            return this;
        }

        public JoinCommunityRequest build() {
            JoinCommunityRequest joinCommunityRequest = new JoinCommunityRequest();
            joinCommunityRequest.setAcceptedAt(acceptedAt);
            joinCommunityRequest.setRejectAt(rejectAt);
            joinCommunityRequest.member = this.member;
            joinCommunityRequest.id = this.id;
            joinCommunityRequest.community = this.community;
            joinCommunityRequest.createdAt = this.createdAt;
            joinCommunityRequest.referrer = this.referrer;
            return joinCommunityRequest;
        }
    }
}

File: src/main/kotlin/co/gatedaccess/web/data/model/JoinCommunityRequest.kt
package co.gatedaccess.web.data.model

import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.DBRef
import org.springframework.data.mongodb.core.mapping.Document
import org.springframework.data.mongodb.core.mapping.Field
import java.util.*

@Document("join_community_request")
class JoinCommunityRequest {
    @Id
    var id: String? = null

    @DBRef
    var member: Member? = null

    @DBRef
    var community: Community? = null

    @Field("created_at")
    @CreatedDate
    var createdAt: Date? = null

    @Field("accepted_at")
    var acceptedAt: Date? = null

    @Field("rejected_at")
    var rejectedAt: Date? = null

    @DBRef
    var referrer: Member? = null

    class Builder {
        private var id: String? = null
        private var member: Member? = null
        private var community: Community? = null
        private var createdAt: Date? = null
        private var acceptedAt: Date? = null
        private var rejectAt: Date? = null
        private var referrer: Member? = null

        fun withId(id: String?): Builder {
            this.id = id
            return this
        }

        fun withMember(member: Member?): Builder {
            this.member = member
            return this
        }

        fun withCommunity(community: Community?): Builder {
            this.community = community
            return this
        }

        fun withCreatedAt(createdAt: Date?): Builder {
            this.createdAt = createdAt
            return this
        }

        fun withAcceptedAt(acceptedAt: Date?): Builder {
            this.acceptedAt = acceptedAt
            return this
        }

        fun withRejectAt(rejectAt: Date?): Builder {
            this.rejectAt = rejectAt
            return this
        }

        fun withReferrer(referrer: Member?): Builder {
            this.referrer = referrer
            return this
        }

        fun build(): JoinCommunityRequest {
            val joinCommunityRequest = JoinCommunityRequest()
            joinCommunityRequest.acceptedAt = acceptedAt
            joinCommunityRequest.rejectedAt = rejectAt
            joinCommunityRequest.member = this.member
            joinCommunityRequest.id = this.id
            joinCommunityRequest.community = this.community
            joinCommunityRequest.createdAt = this.createdAt
            joinCommunityRequest.referrer = this.referrer
            return joinCommunityRequest
        }

        companion object {
            fun aJoinCommunityRequest(): Builder {
                return Builder()
            }
        }
    }
}
